client/src/main/java/io/pravega/client/stream/impl/TxnKeepAlive.java

```diff
@@ -0,0 +1,115 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.client.stream.impl;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import io.pravega.client.stream.EventWriterConfig;
+import io.pravega.client.stream.Stream;
+import io.pravega.common.concurrent.Futures;
+import javax.annotation.concurrent.GuardedBy;
+import lombok.Synchronized;
+import lombok.extern.slf4j.Slf4j;
+
+import static io.pravega.common.concurrent.Futures.getException;
+
+/**
+ * TxnKeepAlive is used to send KeepAlives to renew the transaction lease for active transactions.
+ * It invokes io.pravega.client.stream.impl.Controller#pingTransaction() on the controller to renew the lease of active
+ * transactions. Incase of a controller instance not being reachable the controller client takes care of retrying on a
+ * different controller instance see io.pravega.client.stream.impl.ControllerResolverFactory for details.
+ */
+@Slf4j
+public class TxnKeepAlive implements AutoCloseable {
```

Comment:
The name of this class is a bit misleading. An instance of the this class is not really an instance of a txn keep alive. This is an engine that sends txn keep alive. A few suggestions for the name of this class:

- `Pinger`
- `PingEngine`
- `PingGenerator`
- `TxnLeaseRenewer`

Reply:
Updated the class name to `Pinger`